// @curiousdev-oss/eslint-plugin-web-perf/dist/src/rules/no-heavy-namespace-imports.js
const rule = {
    meta: {
        type: "problem",
        docs: {
            description: "Disallow heavy namespace imports like `import * as _ from 'lodash'`",
        },
        schema: [
            {
                type: "object",
                properties: { deny: { type: "array", items: { type: "string" } } },
                additionalProperties: false,
            },
        ],
    },
    create(context) {
        const options = context.options?.[0] || {};
        const { deny = ["lodash", "moment"] } = options;
        return {
            ImportDeclaration(node) {
                const source = node.source?.value;
                if (!source)
                    return;
                const isDenied = deny.some((pkg) => source === pkg);
                if (!isDenied)
                    return;
                const ns = node.specifiers?.find((s) => s.type === "ImportNamespaceSpecifier");
                if (ns) {
                    context.report({
                        node,
                        message: `Avoid namespace import from '${source}'. Import specific modules or use lighter alternatives.`,
                    });
                }
            },
        };
    },
};
export default rule;
